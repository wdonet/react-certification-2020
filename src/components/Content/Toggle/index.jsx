import React from 'react';

const Toggle = () => {
  return (
    <div>
      <label className="switch">
        <input type="checkbox" />
        <span className="slider"></span>
      </label>
    </div>
  );
};

export default Toggle;
