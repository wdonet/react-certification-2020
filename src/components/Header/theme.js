const base = {
  easeOutBack: 'cubic-bezier(0.34, 1.56, 0.64, 1)',
  colorWhite: 'rgb(255, 255, 255)',
  colorBlack: 'rgb(0, 0, 0)',
};

const dark = {
  id: 'dark',
  ...base,
  backgroundColor: '#333',
  textColor: 'black',
};

const light = {
  id: 'light',
  ...base,
  backgroundColor: '#666',
  textColor: 'white',
};

export const theme = { dark, light };
