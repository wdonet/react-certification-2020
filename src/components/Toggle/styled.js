import styled from 'styled-components';

const Slider = styled.span`
  border-radius: 34px;
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: ${({ theme }) => theme.font.primary};
  -webkit-transition: 0.4s;
  transition: 0.4s;
  &:before {
    border-radius: 50%;
    position: absolute;
    content: '';
    height: 13px;
    width: 13px;
    left: 2px;
    bottom: 1.5px;
    background-color: white;
    -webkit-transition: 0.4s;
    transition: 0.4s;
  }
`;

const ToogleInput = styled.input`
  &:before {
    position: absolute;
    content: '';
    height: 13px;
    width: 13px;
    left: 2px;
    bottom: 1.5px;
    background-color: white;
    -webkit-transition: 0.4s;
    transition: 0.4s;
  }
  &:checked + ${Slider} {
    background-color: ${({ theme }) => theme.font.highlight};
  }
  &:checked + ${Slider}:before {
    -webkit-transform: translateX(13px);
    -ms-transform: translateX(13px);
    transform: translateX(13px);
  }
  &:focus + ${Slider} {
    box-shadow: 0 0 1px ${({ theme }) => theme.font.highlight};
  }
`;

const StyledLabel = styled.label`
  position: relative;
  display: inline-block;
  width: 30px;
  height: 16px;
  ${ToogleInput} {
    opacity: 0;
    width: 0;
    height: 0;
  }
`;

export { ToogleInput, StyledLabel, Slider };
