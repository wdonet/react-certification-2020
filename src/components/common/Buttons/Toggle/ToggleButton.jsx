// React
import React from 'react';
// PropTypes
import PropTypes from 'prop-types';
// Styles
import {
  ToggleWrapper,
  SwitchHandleWrapper,
  Handle,
  SwitchWrapper,
  CheckBox,
  Circle,
} from './styles';

const ToggleButton = ({ checked, handleClick, title }) => {
  return (
    <ToggleWrapper data-testid="toggle-button">
      <SwitchHandleWrapper>
        <Handle>
          <SwitchWrapper>
            <CheckBox checked={checked} onChange={handleClick} />
            <Circle checked={checked} />
          </SwitchWrapper>
        </Handle>
      </SwitchHandleWrapper>
      {title}
    </ToggleWrapper>
  );
};

ToggleButton.propTypes = {
  checked: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  handleClick: PropTypes.func.isRequired,
};

export default ToggleButton;
