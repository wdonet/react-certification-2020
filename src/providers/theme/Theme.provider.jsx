// React
import React, { useState, useEffect, useContext, useCallback } from 'react';
// Utils
import { THEME_SELECTED } from '../../utils/constants';
import { storage } from '../../utils/storage';
// Themes
import light from '../../theme/light';
import dark from '../../theme/dark';

const ThemeContext = React.createContext({ theme: light, selected: false });

const useTheme = () => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error(`Can't use "useTheme" without an ThemeProvider!`);
  }
  return context;
};

const ThemeBaseProvider = ({ children }) => {
  const [theme, setTheme] = useState(light);
  const [selected, setSelected] = useState(false);

  useEffect(() => {
    const lastTheme = storage.get(THEME_SELECTED);
    const whichTheme = lastTheme ? dark : light;
    setSelected(lastTheme);
    setTheme(whichTheme);
  }, []);

  const toggleTheme = useCallback((toggle) => {
    setTheme(toggle ? dark : light);
    storage.set(THEME_SELECTED, toggle);
  }, []);

  return (
    <ThemeContext.Provider value={{ theme, selected, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

export { useTheme, ThemeContext };
export default ThemeBaseProvider;
