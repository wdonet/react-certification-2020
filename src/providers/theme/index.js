export { default, useTheme } from './Theme.provider';
