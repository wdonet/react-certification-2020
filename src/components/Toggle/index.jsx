import React from 'react';
import { StyledLabel, ToogleInput, Slider } from './styled';

const Toggle = ({ onChange }) => {
  return (
    <StyledLabel>
      <ToogleInput onChange={onChange} type="checkbox" />
      <Slider />
    </StyledLabel>
  );
};

export default Toggle;
