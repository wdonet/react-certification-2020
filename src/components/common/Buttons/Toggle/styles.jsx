import styled from 'styled-components';

export const ToggleWrapper = styled.label`
  cursor: pointer;
  display: inline-flex;
  align-items: center;
  vertical-align: middle;
  margin-right: 16px;
`;

export const SwitchHandleWrapper = styled.span`
  width: 58px;
  height: 38px;
  display: inline-flex;
  padding: 12px;
  z-index: 0;
  overflow: hidden;
  box-sizing: border-box;
  flex-shrink: 0;
  vertical-align: middle;
`;

export const SwitchWrapper = styled.span`
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-start;
`;

export const CheckBox = styled.input.attrs({ type: 'checkbox' })`
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;
`;

export const Circle = styled.span`
  width: 20px;
  height: 20px;
  box-shadow: 0px 2px 1px -1px rgb(0 0 0 / 20%), 0px 1px 1px 0px rgb(0 0 0 / 14%),
    0px 1px 3px 0px rgb(0 0 0 / 12%);
  border-radius: 50%;
  background-color: ${(props) => props.theme.palette && props.theme.palette.accent};
  transform: translateX(${(props) => (props.checked ? '100%' : '0')});
  transition: ease 0.3s;
`;

export const Handle = styled.span`
  width: 100%;
  height: 100%;
  z-index: -1;
  transition: opacity 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 7px;
  background-color: rgba(0, 0, 0, 0.38);
`;
