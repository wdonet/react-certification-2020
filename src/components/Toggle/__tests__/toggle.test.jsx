import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';

import Toggle from '../index';

import { lightTheme } from '../../../theme';

describe('Toggle', () => {
  test('renders correctly', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={lightTheme}>
          <Toggle />
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
