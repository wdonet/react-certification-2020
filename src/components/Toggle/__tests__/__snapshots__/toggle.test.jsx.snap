// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders correctly 1`] = `
.c4 {
  border-radius: 34px;
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #333333;
  -webkit-transition: 0.4s;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.c4:before {
  border-radius: 50%;
  position: absolute;
  content: '';
  height: 13px;
  width: 13px;
  left: 2px;
  bottom: 1.5px;
  background-color: white;
  -webkit-transition: 0.4s;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.c2:before {
  position: absolute;
  content: '';
  height: 13px;
  width: 13px;
  left: 2px;
  bottom: 1.5px;
  background-color: white;
  -webkit-transition: 0.4s;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.c2:checked + .c3 {
  background-color: #eea2a2;
}

.c2:checked + .c3:before {
  -webkit-transform: translateX(13px);
  -ms-transform: translateX(13px);
  -webkit-transform: translateX(13px);
  -ms-transform: translateX(13px);
  transform: translateX(13px);
}

.c2:focus + .c3 {
  box-shadow: 0 0 1px #eea2a2;
}

.c0 {
  position: relative;
  display: inline-block;
  width: 30px;
  height: 16px;
}

.c0 .c1 {
  opacity: 0;
  width: 0;
  height: 0;
}

<label
  className="c0"
>
  <input
    className="c1 c2"
    type="checkbox"
  />
  <span
    className="c3 c4"
  />
</label>
`;
